### ISSUE TITLE
Missing native menu items on macOS

### ISSUE THREAD

**Issue by msanders** [alacritty/alacritty#3245]
The lack of menu items on macOS make Alacritty feel slightly out of place / non-native. Currently, there are no menu items, not even Quit. It seems there should at least be the following:

- Alacritty => About, Services, Hide/Show, Quit
- Edit (defaults)
- Window (defaults)
- Help (defaults, with link pointing to online docs)

I've never seen an application without these, except for Calculator which is missing Services.

This should be the default, but perhaps it could be configurable as well.

**Comment by nixpulvis** [comment]
I'd ague we should only do the minimal work here. On Linux and Windows I think we can avoid these altogether. So on macOS we should maintain only the most expected entries. I'd say the "issue" here is mostly that macOS always draws an dropdown for "Alacritty" (app name) which we currently don't fill with any entries, which is a bit startling. At the very least I'd expect a "Quit" option. A help link would be nice, but completely optional IMO.

More information can be found here: https://developer.apple.com/design/human-interface-guidelines/macos/menus/menu-bar-menus/#app-menu

**Comment by msanders** [comment]
Agree with the minimalism, but IMO the default macOS-only keybindings should all correspond to menu items, and optionally be turned off if you want a super lean setup. Basically, it should look/behave like other native apps by default.

Help is nice to have since it's included in every other app I've seen, and it has a built-in search function for other menu items.

**Comment by chrisduerr** [comment]
> At the very least I'd expect a "Quit" option

Why? Is this documented somewhere in a guideline from macOS that this should be possible?

If macOS doesn't do it by default, that seems mostly like a place to put arbitrary application defined shortcuts. So I don't see a reason why we would copy other unrelated applications here?

What's the concrete benefit to this, if it is not a standard way to close applications supported by macOS anyways?

**Comment by msanders** [comment]
> If macOS doesn't do it by default, that seems mostly like a place to put arbitrary application defined shortcuts. So I don't see a reason why we would copy other unrelated applications here?

The items listed above are effectively system defaults and cohesive across all applications I've tried on my machine. They come from the default entries in a new interface builder file.

> What's the concrete benefit to this, if it is not a standard way to close applications supported by macOS anyways?

It is standard. See the above guidelines linked above in [Apple’s HIG](https://developer.apple.com/design/human-interface-guidelines/macos/menus/menu-bar-menus/):

> **Implement standard menus and submenus.** People generally expect most standard menus and menu items to exist unless they’re irrelevant to your app. In many cases, the behavior and functionality of standard menu items is automatically implemented for you. For example, when the user selects text in a standard text field, the Edit > Copy menu item is made available for use.
>
> **Enable the expected keyboard shortcuts for standard menu items.** Keyboard shortcuts for standard menu items—like Copy, Cut, Paste, Save, and Print—should be consistent in all apps.

I'd be interested in taking a stab at submitting a contribution for this if it's something other folks would like.

**Comment by chrisduerr** [comment]
> I'd be interested in taking a stab at submitting a contribution for this if it's something other folks would like.

I'm not opposed to the idea, assuming we keep it simple.

Stuff like closing the window makes a lot of sense, but I don't think it's necessary to have a menu entry for every default macOS binding for example.

**Comment by daneharrigan** [comment]
I hadn't noticed the lack of menu items until seeing this issue. I agree with the reasoning on keeping the menu as minimal as possible. Having an "Alacritty > Quit Alacritty" menu option seems reasonable. I'd even prefer for the entire menu to not exist, if it was possible.

**Comment by timw4mail** [comment]
`About` and `Quit` are about the only menu items I'd recommend (in the `Alacritty` top level menu. `About` would simply display the logo and version, which is helpful for knowing what version of Alacritty is running.

Kitty has this kind of `About` window:
![Screen Shot 2020-03-18 at 10 46 43 AM](https://user-images.githubusercontent.com/741525/76973101-dd7e0d00-6905-11ea-9779-9dafea5f2513.png)

It's a bit weird not to see a top-level `Help` menu, but I'm not sure I've ever seen that menu be useful either.

**Comment by cole-h** [comment]
If you just want to see the version, why not just run `alacritty -V` or `alacritty --version`?

**Comment by peterkos** [comment]
The first thing I personally tried to do was `Cmd+,` to change a preference, which didn't work. 
As a Mac user I expect the basics:
- `Cmd+,` to open preferences, either a file (as in Sublime Text) or window
- `Cmd+n` to open a new window
- `Cmd+w` to close the current window (works! but no menu item)
- `Cmd+q` to quit the app (works! but no menu item)
- `Cmd++` and `Cmd+-` to zoom (works! but no menu item)
etc.

I think a good rule of thumb is any common keyboard shortcut should have a menu item. Maybe look to the default Terminal app to see what they put there? Or Sublime Text, as they also have a super speed/keyboard-focused architecture.

**Comment by nixpulvis** [comment]
@peterkos 

1. No terminal emulator is going to open sublime text for you, use `subl ~/.config/alacritty/alacritty.yml` if you want to open it in that application
2. See #607

**Comment by peterkos** [comment]
@nixpulvis Yes, I know how to open files using Sublime's terminal command -- I mean somehow opening the ability to change preferences via a file through the user interface as a potential option -- somehow triggering `$EDITOR /.config/alacritty`; something along those lines. Or having a GUI for it, like every other Mac application.
(I know this might go against the idea of "super fast" but it _does_ run alongside other GUI apps.)

And oops, I didn't realize that was an open issue already, or that it wasn't yet implemented. I was just referring to a list of common actions users expect from a GUI app, that are available in the menu bar, to follow OS convention.

**Comment by kchibisov** [comment]
if you really want that just add a binding to alacritty to open your config with the editor you want. Consult our keybinding section on how to do that, basically you want to launch a program with args, on a key binding.

**Comment by chrisduerr** [comment]
> I mean somehow opening the ability to change preferences via a file through the user interface as a potential option -- somehow triggering $EDITOR /.config/alacritty; something along those lines. Or having a GUI for it, like every other Mac application.

This won't happen.

**Comment by dancojocaru2000** [comment]
> This won't happen.

Why?

**Comment by daneharrigan** [comment]
Let's lock the issue and close it.

**Comment by dancojocaru2000** [comment]
When macOS will have support for windows (#607), native menu items will be a natural continuation. If macOS windows will be implemented properly, Alacritty will remain running despite having no windows open, and then `Command+Tab`-ing and executing keyboard shortcuts associated to menu items is a common pattern. For example, Cmd+Tab to Alacritty and then Cmd+Q to when no windows are open.

Bringing up the emoji / symbols popup (#1895) is also generally achieved by having an Edit > Emoji & Symbols menu with an associated `Cmd+Ctrl+Space` keyboard shortcut.

**Comment by jackno** [comment]
Also missing App Shortcuts. The ones you can set in `System Preferences` -> `Keyboard` -> `Shortcuts` -> `App Shortcuts`. I'm not sure if these are the same as menu items.

<img width="744" alt="Screen Shot 2020-05-21 at 10 07 35 PM" src="https://user-images.githubusercontent.com/8162609/82627208-7b64b100-9baf-11ea-9592-903d1f515119.png">

**Comment by dancojocaru2000** [comment]
Inb4 "just use the app config file for shortcuts" instead of using the OS way. A solution, sure, but an inconvenient.

**Comment by piksel** [comment]
There is also apps like [CheatSheet](https://mediaatelier.com/CheatSheet/) that displays the keyboard bindings for the current app when you hold down the command key. That relies on the menu entries to populate it's fields (and is of course empty for Alacritty right now)

**Comment by droustchev** [comment]
One thing I really miss, which I believe is related to the lack of native menu, is the ability to launch a new instance of Alacritty via the key-binding when it is active, but not on the current space. Whenever I need a separate instance, I have to find the space with my Alacritty window, cmd+n to get a new one and then move it to the space I need it.

Is my understanding correct, that they shortcuts would need to be implemented in the menu bar, which complicates things since Alacritty's keybindings are very very flexible and basically infinite?

**Comment by EdmundsEcho** [comment]
Great app.  Thank you.  I particularly like the ability to create a  minimalist view of my work.  

Alacritty does not have split windows for macos; the priority not to do so makes sense.

However, per  @hexcowboy observation,  it also does not support the macos keyboard shortcut that every macos app uses.  This *would have been* a quick way to mollify the need for a split window by instead using the macos "side-by-side" window feature.  

The good news is that Alacritty does have this "side-by-side" capacity when the window buttons are present - *a mouse click* away.  However, use of the mouse is not ideal.

I don't know the answer: is it a lot more effort  to trigger the behavior already in place using the keyboard shortcut?  How might take the first step to assessing the effort?

**Comment by Neob91** [comment]
I'm not sure this is related to this issue, but I'm using Alacritty full-screen, and I'm unable to even see the MacOS system bar when I move my mouse to the top (which prevents me from inspecting the current time/date, viewing Wi-Fi/Bluetooth settings, battery, etc.)

Anyway, aside from that, seems to work great!

**Comment by BogdanTheGeek** [comment]
Going down the rabbit hole, this "issue" or missing feature is because alacritty doesn't expose the window submenu.
Apparently, all one would need to do to allow this is inheriting the system NSMenu.
[Wezterm](https://github.com/wez/wezterm) implements this.
I tried porting it myself, but I not very familiar with rust nor AppKit.
[This](https://forum.xojo.com/t/let-the-macos-manage-your-window-menu/64232) seems to show the ObjC way of handling the window menu.

**Comment by samxiao-bedrock-analytics** [comment]
> Going down the rabbit hole, this "issue" or missing feature is because alacritty doesn't expose the window submenu. Apparently, all one would need to do to allow this is inheriting the system NSMenu. [Wezterm](https://github.com/wez/wezterm) implements this. I tried porting it myself, but I not very familiar with rust nor AppKit. [This](https://forum.xojo.com/t/let-the-macos-manage-your-window-menu/64232) seems to show the ObjC way of handling the window menu.

Agreed! I am not sure why Alacritty don't implement this, but still waiting for this feature in 2025!

**Comment by kocoai** [comment]
There is [Ghostty](https://github.com/ghostty-org/ghostty) with different approach: **platform-native UI**, which is lot more user friendly for macOS (and other OS)

**Comment by marcingrabda** [comment]
@kocoai Great news. Time to ditch Alacritty.

**Comment by NicolaiKuertoes** [comment]
> [@nixpulvis](https://github.com/nixpulvis) Yes, I know how to open files using Sublime's terminal command -- I mean somehow opening the ability to change preferences via a file through the user interface as a potential option -- somehow triggering `$EDITOR /.config/alacritty`; something along those lines. Or having a GUI for it, like every other Mac application. (I know this might go against the idea of "super fast" but it _does_ run alongside other GUI apps.)
> 
> And oops, I didn't realize that was an open issue already, or that it wasn't yet implemented. I was just referring to a list of common actions users expect from a GUI app, that are available in the menu bar, to follow OS convention.
> 
> Edit: I mention Sublime because when you press`Cmd ,`, it opens the raw config file instead of a normal preference GUI. I assume this project would prefer that approach?

```toml
[[keyboard.bindings]]
key = "Comma"
mods = "Command"
command = { program = "open", args = ["-a", "Alacritty", "-n", "--args", "-e", "vim", "/Users/<USERNAME>/.config/alacritty/alacritty.toml"] }
```
